### PR TITLE
Dead letter exchange

### PR DESCRIPTION
**Pull request by felix-zg**

### CODE REVIEW

**Comment by felix-zg** [comment]
image

File: index.js [hunk 1]
```diff
@@ -53,15 +77,22 @@ module.exports = class RabbitMQ extends EventEmitter {
...
                 await handler(JSON.parse(msg.content.toString('utf8')));
                 await this.channel.ack(msg);
             } catch (err) {
                 log.error({err, msg: msg.content.toString('utf8')});
-                if (err instanceof SyntaxError) await this.channel.ack(msg);
-                else await this.channel.nack(msg);
+
+                if (!this.exchange || err instanceof SyntaxError || msg.properties?.headers['x-death']?.[0]?.count >= SEND_TO_DLQ_AFTER) {
+                    await this.channel.assertQueue(DLQ_NAME);
+                    await this.channel.publish('', DLQ_NAME, Buffer.from(JSON.stringify(msg)));
```
Comment (gilad-deel) [inline index.js:91]: replace content with `content.toString('utf8')`

File: index.js [hunk 2]
```diff
@@ -13,7 +20,19 @@ module.exports = class RabbitMQ extends EventEmitter {
...
+                            await this.channel.assertExchange(`${this.exchange}_${RYX_NAME}`, 'fanout');
+                            await this.channel.assertQueue(`${this.exchange}_${RYQ_NAME}`, {
+                                arguments: {
+                                    'x-dead-letter-exchange': this.exchange,
+                                    'x-message-ttl': RETRY_TIMEOUT,
+                                },
+                            });
+                            await this.channel.bindQueue(`${this.exchange}_${RYQ_NAME}`, `${this.exchange}_${RYX_NAME}`, '#');
+
+                            await this.channel.assertExchange(this.exchange, 'x-delayed-message', {arguments: {'x-delayed-type': 'direct'}});
```
Comment (gilad-deel) [inline index.js:34]: make this first in block

File: index.js [hunk 3]
```diff
@@ -40,7 +59,12 @@ module.exports = class RabbitMQ extends EventEmitter {
 
     async assertQueue(queue) {
         await this.connection;
-        if (!this.queues[queue]) this.queues[queue] = this.channel.assertQueue(queue);
+        if (!this.queues[queue])
+            this.queues[queue] = this.channel.assertQueue(queue, {
+                arguments: {
+                    'x-dead-letter-exchange': `${this.exchange}_${RYX_NAME}`,
+                },
+            });
```
Comment (gilad-deel) [inline index.js:67]: 1 line

**Review by gilad-deel — CHANGES_REQUESTED** [review]

**Comment by felix-zg** [comment]
`{"fields":{"consumerTag":"amq.ctag-SwkySIO4h8gcVqWVPMsv8w","deliveryTag":501,"redelivered":false,"exchange":"/f","routingKey":"/f"},"properties":{"headers":{"x-death":[{"count":5,"exchange":"/f_RetryExchange","queue":"/f_RetryQueue","reason":"expired","routing-keys":["/f"],"time":{"!":"timestamp","value":1631177824}},{"count":5,"exchange":"/f","queue":"/f","reason":"rejected","routing-keys":["/f"],"time":{"!":"timestamp","value":1631177819}}],"x-first-death-exchange":"/f","x-first-death-queue":"/f","x-first-death-reason":"rejected"}},"content":"\"{\\\"test\\\":0}\""}
`

**Review by gilad-deel — APPROVED** [review]

**Review by gilad-deel — APPROVED** [review]

**Pull request merged**